## Hacker News Posts project
- the dataset is from a website where users submit posts and they can be commented
- the project focus on taking a closer look at the post where the user ask a question to the webiste Hacker News: **Ask HN**. 
- the other interests is to analayse the posts where the user show a project to the website Hacker News: **Show HN**

**The questions:**
- *out of the 2, what type gets more comments on average?* 
- *does the time of the post influences the number of posts in average?*

we import the dataset as a list

In [4]:
from csv import reader
file=open('HN_posts_year_to_Sep_26_2016.csv')
read_file=reader(file)
hn=list(read_file)

In [3]:
read_file

In [59]:
hn[0:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]



- splitting of the dataset in headers and the data itself

In [60]:
headers=hn[0] #extract headers
hn=hn[1:] #remove headers from dataset

In [61]:
headers

hn[0:5]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

- dectection of the post starting  with either **Ask HN** or **Show HN**  

In [62]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title=row[1]
    title=title.lower()
    if title.startswith('ask'):
        ask_posts.append(row)
    elif title.startswith('show'):
        show_posts.append(row)
    else:
        other_posts.append(row)
print(len(hn))
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))
print(len(ask_posts+show_posts+other_posts))

20100
1756
1164
17180
20100


it we seems to have split correctly the news without losing any posts.

In [63]:
show_posts[0:5]

[['10627194',
  'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform',
  'https://iot.seeed.cc',
  '26',
  '22',
  'kfihihc',
  '11/25/2015 14:03'],
 ['10646440',
  'Show HN: Something pointless I made',
  'http://dn.ht/picklecat/',
  '747',
  '102',
  'dhotson',
  '11/29/2015 22:46'],
 ['11590768',
  'Show HN: Shanhu.io, a programming playground powered by e8vm',
  'https://shanhu.io',
  '1',
  '1',
  'h8liu',
  '4/28/2016 18:05'],
 ['12178806',
  'Show HN: Webscope  Easy way for web developers to communicate with Clients',
  'http://webscopeapp.com',
  '3',
  '3',
  'fastbrick',
  '7/28/2016 7:11'],
 ['10872799',
  'Show HN: GeoScreenshot  Easily test Geo-IP based web pages',
  'https://www.geoscreenshot.com/',
  '1',
  '9',
  'kpsychwave',
  '1/9/2016 20:45']]

- time to check what type of post get the max average number of posts

loop into each respective dataset and summing of each comments

In [64]:
#average numbers post for ask posts
total_ask_comments=0
for row in ask_posts:
    num_comments=row[4]
    num_comments=int(num_comments)
    total_ask_comments += num_comments

avg_ask_comments=total_ask_comments/len(ask_posts)
print(avg_ask_comments)

14.203302961275627


In [65]:
#average numbers pot for show posts
total_show_comments=0
for row in show_posts:
    num_comments=row[4]
    num_comments=int(num_comments)
    total_show_comments += num_comments
avg_show_comments=total_show_comments/len(show_posts)
print(avg_show_comments)

10.323024054982818


**conclusion** 

Ask HN posts get in average 4 comments more than Show HN post (14.2 vs 10.32 respectively).

For the rest of the study, only Ask HN are considered.

- does the time of publication matters for the numbers of comments and the number of posts??

In [66]:
print(headers)
ask_posts[-1][-1]

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


'5/21/2016 9:22'

the date fortmat is month/day/year and time is 24hours

In [67]:
for row in ask_posts:
    print(row[-1])

8/16/2016 9:55
11/22/2015 13:43
5/2/2016 10:14
8/2/2016 14:20
10/15/2015 16:38
9/26/2015 23:23
4/22/2016 12:24
11/16/2015 9:22
2/24/2016 17:57
6/4/2016 17:17
9/19/2015 17:04
9/22/2015 13:16
6/21/2016 15:45
1/13/2016 21:17
10/4/2015 21:27
1/25/2016 20:27
10/27/2015 2:47
1/19/2016 12:01
3/22/2016 2:05
9/8/2015 14:04
8/28/2016 18:06
7/20/2016 13:44
9/12/2016 16:52
2/29/2016 17:52
4/18/2016 15:28
12/28/2015 14:38
4/4/2016 3:34
1/15/2016 21:47
11/19/2015 5:33
12/20/2015 3:59
10/15/2015 21:34
2/26/2016 19:20
8/2/2016 18:00
2/28/2016 1:24
1/13/2016 9:12
5/6/2016 1:14
6/23/2016 13:59
4/30/2016 17:21
10/20/2015 19:21
10/25/2015 15:09
5/4/2016 14:14
12/23/2015 20:48
6/1/2016 16:19
7/10/2016 22:19
6/21/2016 23:40
4/24/2016 19:36
5/20/2016 1:26
6/24/2016 17:02
7/12/2016 19:34
10/16/2015 16:36
3/15/2016 22:15
11/24/2015 1:05
5/30/2016 8:14
1/11/2016 4:53
8/28/2016 20:38
3/25/2016 12:52
4/14/2016 0:58
12/31/2015 14:18
2/26/2016 10:23
4/1/2016 16:24
6/22/2016 21:15
10/13/2015 17:07
3/3/2016 10:40
4/3

In [68]:
import datetime as dt

result_lists=[] # list of list containing the time of publication and number of comments for each posts in ask questions

for row in ask_posts:
    time=row[6]
    n_comments=int(row[4])
    post=[time, n_comments]
    result_lists.append(post)

In [69]:
print(type(result_lists[0][0]))

<class 'str'>


for the moment, the date is still consider as a string, the datetime.strptime() methods will be employed to format the time and isolates the hour info

In [70]:
counts_by_hours={} #number of posts per hour
comments_by_hours={} #number of comments per hour
for row in result_lists:
    #isolation of the hour data
    time_str=row[0]
    time_dt=dt.datetime.strptime(time_str,'%m/%d/%Y %H:%M')
    hour=time_dt.strftime("%H")
    #number of comments
    n_comments=row[1]
    
    #number of post per hour
    if hour in counts_by_hours:
        counts_by_hours[hour] +=1
    else:
        counts_by_hours[hour]=1
    
    #number of comments per hour
    if hour in comments_by_hours:
        comments_by_hours[hour] += n_comments
    else:
        comments_by_hours[hour] = n_comments 
    


In [71]:
counts_by_hours

{'00': 55,
 '01': 60,
 '02': 58,
 '03': 55,
 '04': 48,
 '05': 46,
 '06': 44,
 '07': 34,
 '08': 48,
 '09': 45,
 '10': 60,
 '11': 58,
 '12': 73,
 '13': 86,
 '14': 110,
 '15': 116,
 '16': 109,
 '17': 100,
 '18': 109,
 '19': 112,
 '20': 80,
 '21': 109,
 '22': 71,
 '23': 70}

as expecting there are most in the afternoon and the tendancy last until evening when movies starts

In [72]:
comments_by_hours

{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 459,
 '04': 339,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 794,
 '11': 641,
 '12': 687,
 '13': 1254,
 '14': 1420,
 '15': 4477,
 '16': 1949,
 '17': 1146,
 '18': 1439,
 '19': 1295,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 713}

interesting to see that posts published between 3 and 4pm are the one with the highest number of comments. 
*it could be explain as when people come back from work it will be the post to be seen at the top of their feed? the small increase of number of post (see cell above) at 15 does not explain the steep increase of comments for 15.*
       
       this might be not the correct explanation, the hours labeleled are from EST, in Europe is 6 hours more.

the calculation of the **average number of comments per post according to the hours** might help to get a better overview of the trend

In [73]:
len(counts_by_hours
   )

24

In [74]:
avg_by_hour=[]
for hour in counts_by_hours:
    average = comments_by_hours[hour] / counts_by_hours[hour]
    avg_by_hour.append([hour, average])
avg_by_hour

[['09', 5.5777777777777775],
 ['03', 8.345454545454546],
 ['16', 17.880733944954127],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['15', 38.5948275862069],
 ['11', 11.051724137931034],
 ['05', 10.08695652173913],
 ['08', 10.25],
 ['12', 9.41095890410959],
 ['10', 13.233333333333333],
 ['22', 6.746478873239437],
 ['17', 11.46],
 ['00', 8.127272727272727],
 ['02', 23.810344827586206],
 ['04', 7.0625],
 ['20', 21.525],
 ['19', 11.5625],
 ['14', 12.909090909090908],
 ['13', 14.581395348837209],
 ['23', 10.185714285714285],
 ['18', 13.20183486238532],
 ['21', 16.009174311926607],
 ['01', 11.383333333333333]]

In [75]:
swap_avg_by_hour=[]
for hour in avg_by_hour:
    swap_avg_by_hour.append([hour[1],hour[0]])
max_avg=max(swap_avg_by_hour)
print('The best time to publish is a post is {0} with {1:.2f} comments per post in average'.format(max_avg[1],max_avg[0]))

The best time to publish is a post is 15 with 38.59 comments per post in average


In [79]:
sorted_swap=sorted(swap_avg_by_hour,reverse=True)

In [81]:
sorted_swap[0:5]

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [17.880733944954127, '16'],
 [16.009174311926607, '21']]

In [97]:
import datetime as dt
for row in sorted_swap[:5]:
    avg=row[0]
    time=row[1]
    time=dt.datetime.strptime(time,'%H')+dt.timedelta(hours=6)
    time=dt.datetime.strftime(time,'%H:%M:')
    
    print('at {} you will get {:.2f} average comments per post'.format(time,avg))
    
    

at 21:00: you will get 38.59 average comments per post
at 08:00: you will get 23.81 average comments per post
at 02:00: you will get 21.52 average comments per post
at 22:00: you will get 17.88 average comments per post
at 03:00: you will get 16.01 average comments per post


## going further: 
Determine if show or ask posts receive more points on average.
Determine if posts created at a certain time are more likely to receive more points.
Compare your results to the average number of comments and points other posts receive.
